### PR TITLE
fix: Sending envelopes multiple times

### PR DESCRIPTION
**Pull request by philipphofmann**
## :scroll: Description

Sending all cached envelopes in SentryHttpTransport had a concurrency issue. For each envelope to send SentryHttpTransport iterates over all cached envelopes and queues a request. The envelopes get removed from the cache once a request finishes with success. If multiple envelopes get sent in a short time interval SentryHttpTransport iterates over cached envelopes for which already a request is queued. This can lead to multiple requests for the same envelope and other unpredictable behaviour as the session stats generated by SentrySessionGeneratorTests being random on Sentry. This issue is fixed now by sequentially sending cached envelopes. When one envelope is sent successfully the next envelope in the cache is queued for sending. If SentryHttpTransport is already sending envelopes and it receives a new envelope the envelope is just stored in the cache and sending all cached envelopes is not triggered again because it is already running. The new cached envelope is going to be picked up later.

## :bulb: Motivation and Context

See description.

## :green_heart: How did you test it?
With unit tests and SentrySessionGeneratorTests.

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify the changes
- [x] I've updated the CHANGELOG
- [x] No breaking changes

### CODE REVIEW

File: Sources/Sentry/SentryHttpTransport.m [hunk 1]
```diff
@@ -84,36 +92,47 @@ - (void)sendEnvelope:(SentryEnvelope *)envelope
...
+    NSURLRequest *request = [self createEnvelopeRequest:rateLimitedEnvelope
+                                       didFailWithError:requestError];
+
+    if (nil != requestError) {
+        [SentryLog logWithMessage:requestError.localizedDescription andLevel:kSentryLogLevelError];
+        [self.fileManager removeFileAtPath:oldestEnvelope.path];
+        [self sendNextCachedEnvelope];
+        return;
+    } else {
+        [self sendCached:request withFilePath:oldestEnvelope.path];
```
Comment (bruno-garcia) [inline Sources/Sentry/SentryHttpTransport.m:135]: This seem to be the only use of `sendCache`, is that right? if so, can we inline to avoid having more private methods accessing the flag under a lock?

File: Sources/Sentry/SentryHttpTransport.m [hunk 2]
```diff
@@ -139,6 +166,12 @@ - (void)sendCached:(NSURLRequest *)request withFilePath:(NSString *)filePath
             if (nil != response) {
                 [_self.fileManager removeFileAtPath:filePath];
                 [_self.rateLimits update:response];
+
+                [_self sendNextCachedEnvelope];
+            } else {
+                @synchronized(_self) {
+                    _self.isSending = NO;
```
Comment (bruno-garcia) [inline Sources/Sentry/SentryHttpTransport.m:173]: Do we need these private methods? It's super hard to follow the lock and this flag flipping.
I wrote 2 comments only to realize the private method call flipped the flag

File: Sources/Sentry/SentryHttpTransport.m [hunk 3]
```diff
@@ -84,36 +92,47 @@ - (void)sendEnvelope:(SentryEnvelope *)envelope
 // flow
 - (void)sendAllCachedEnvelopes
 {
-    if (![self.requestManager isReady]) {
-        return;
+    @synchronized(self) {
```
Comment (bruno-garcia) [inline Sources/Sentry/SentryHttpTransport.m:95]: This is only called in a background thread, right?

File: Sources/Sentry/SentryHttpTransport.m [hunk 4]
```diff
@@ -19,6 +19,13 @@
 @property (nonatomic, strong) id<SentryRateLimits> rateLimits;
 @property (nonatomic, strong) SentryEnvelopeRateLimit *envelopeRateLimit;
 
+/**
+ * Synching with a dispatch queue to have concurrent reads and writes as barrier blocks is roughly
+ * 30% slower. As this is the only part of this class that needs synchronization it is fine to sync
+ * on self.
+ */
+@property (nonatomic) BOOL isSending;
```
[the remark this answers is not in the document]
Comment (philipphofmann) [reply Sources/Sentry/SentryHttpTransport.m:27]: Good point. I tried with a dispatch queue first, because I was afraid of using `@synchronized` or `atomic` might slow it down too much, but I found out that is actually faster. Basically using `atomic` here is roughly the same as we have right now with all the `@synchronized` keyword. I just found out again "premature optimization is the root of all evil" 😄

File: Sources/Sentry/SentryHttpTransport.m (hunk 3 above)
Comment (philipphofmann) [reply Sources/Sentry/SentryHttpTransport.m:95]: Nope, the first time it is called from `sendEnvelope` and this could also be the main/UI thread.

File: Sources/Sentry/SentryHttpTransport.m (hunk 2 above)
Comment (philipphofmann) [reply Sources/Sentry/SentryHttpTransport.m:173]: I'm going to remove it. 👍

**Comment by philipphofmann** [comment]
> Maybe we can go through this on a call?

I would love that 👍

File: Sources/Sentry/SentryHttpTransport.m (hunk 3 above)
Comment (bruno-garcia) [reply Sources/Sentry/SentryHttpTransport.m]: Task scheduled to follow up on this

File: Sources/Sentry/SentryHttpTransport.m (hunk 4 above)
[the remark this answers is not in the document]
Comment (bruno-garcia) [reply Sources/Sentry/SentryHttpTransport.m]: lol, yeah

Also: 1 I/O equals 100000 calls (on a single thread, take contention away) to a @synchronized method.

And all that is right, or wrong, we don't know, unless we measure.
